### ISSUE TITLE
Dracut network-manager, RHEL 8, and Clevis prevents iface setting changes

### ISSUE THREAD

**Issue by superteece** [dracutdevs/dracut#1062]
I'm using RHEL 8 in a scenario where I provision a device in a DHCP network then later deploy it into a static IP network. My issue occurs when I use nmtui or nmcli to change the interface from dynamic to static, then reboot, it's dynamic again. This seems to be due to the existence of a file in /run/NetworkManager/system-connections which is a characteristic of Dracut's network-manager module that became default over network-legacy in RHEL 8.3.

I see that this use case was doable in network legacy by passing the omit_dracutmodules+=" ifcfg " parameter. Is there an equivalent for network-manager?

There's a related bug report here: https://bugzilla.redhat.com/show_bug.cgi?id=1818579

**Distribution used**
RHEL 8.3

**Dracut version**
Version: 049-95.git20200804.el8

**Init system**
Which init system is being used?

**To Reproduce**
1. Install Clevis, bind to Tang, setup Dracut as the unlocker
2. Reboot to see that LUKS auto unlocks
3. Use nmtui or nmcli to configure a static IP
4. Reboot
5. Host is reverted to DHCP

Something at boot is creating a profile for Wired Connection. If this is deleted or modified with a static address, a new Wired Connection is generated to replace the deleted or modified one at the next boot. If a new profile is created for the static settings, the auto generated Wired Connection is used as default at next boot.

**Expected behavior**
I should be able to change network settings and have them stick? 

**Additional context**
I also posted this issue on the Clevis GitHub as I do not know where a modification should be made or where the solution I'm over looking lies. https://github.com/latchset/clevis/issues/290

**Comment by johannbg** [comment]
@superteece what happens if you boot once and rebuild your initramfs, do configurations "stick" then?

**Comment by johannbg** [comment]
Also what happens if you simply add rd.neednet=1 to the kernel command line?

**Comment by superteece** [comment]
rd.neednet=1 is already set by Clevis

As far as rebuilding initramfs after first boot, I'm not in a spot to try this at the moment. For now, reimaging with RHEL 8.2 followed by adding the omit_dracutmodules+=" ifcfg " parameter and creating a systemd service to flush the interface's IP at boot is resolving the issue. But that's due to RHEL 8.2 using network-legacy.

**Comment by tyll** [comment]
Do I understand correctly that you would like to use DHCP in the initramfs but a static address in the real root?

**Comment by superteece** [comment]
> @superteece Do I understand correctly that you would like to use DHCP in the initramfs but a static address in the real root?

Yes this is accurate. The network on which these devices are provisioned is DHCP due to PXE. After provisioning they are transferred to remote sites which use static addresses. 

However, if initramfs remains DHCP, Dracut panics when an address fails to be assigned.

**Comment by tyll** [comment]
> However, if initramfs remains DHCP, Dracut panics when an address fails to be assigned.

I dont' follow. If you provide kernel arguments for DHCP, then initramfs will use DHCP also in 8.2, regardless of any changes to the profile on disk, wouldn't it?

**Comment by johannbg** [comment]
The panic happens when they are transferred to remote sites which use static addresses as in the dhcp kernel parameter is not removed in the same process as it should be thus dracut picks it up and halts when there is no response to the dhcp request.

**Comment by tyll** [comment]
> The panic happens when they are transferred to remote sites which use static addresses as in the dhcp kernel parameter is not removed in the same process as it should be thus dracut picks it up and halts when there is no response to the dhcp request.

Can you maybe rephrase this? What do you mean with "remote sites" here?

@superteece 
Did you try this with the network-legacy module and omit_dracutmodules+=" ifcfg " or do you just assume that this worked for your use case?

**Comment by johannbg** [comment]
It means as he mentioned it `The network on which these devices are provisioned is DHCP due to PXE. After provisioning they are transferred to remote sites which use static addresses.` ( which I'm probably misunderstanding ). 

What release of dracut is currently in RHEL? The bug report mentions 50 on Fedora ( and probably the underlying cause for was this change ) https://github.com/dracutdevs/dracut/commit/5965710e018989b02a56e8d190b71740ca3b5463#diff-fcc4de97e69e134f38b5bc2cd5222466881d89f9ad189c075dbb20b34d88ee6e )  but there are changes in 51 which might address this 
https://github.com/dracutdevs/dracut/commit/faea4e4ddb10f697590b80f8f17181341c537262#diff-fcc4de97e69e134f38b5bc2cd5222466881d89f9ad189c075dbb20b34d88ee6e
https://github.com/dracutdevs/dracut/commit/3dcaa97ca4dcfa8092252a22df62c60941e59ce3#diff-fcc4de97e69e134f38b5bc2cd5222466881d89f9ad189c075dbb20b34d88ee6e

So this really needs to be tested on more recent dracut release ( 51+ ) to see if this is still an issue.

**Comment by tyll** [comment]
> It means as he mentioned it `The network on which these devices are provisioned is DHCP due to PXE. After provisioning they are transferred to remote sites which use static addresses.` ( which I'm probably misunderstanding ).

Thank you, I understand now. From what I understand, as long as there are dracut/kernel ip command line options, they determine whether the system uses DHCP or static IP. So if the switch from DHCP->static only happens with nmcli but without changing the kernel cmdline, then the system will not boot (if there is no DHCP). If the kernel cmdline is adjusted to static IP, NM should not do any DHCP afterwards. But if the kernel cmdline is static, then there is no need to change from an initrd network config to a different one in the real root.

**Comment by johannbg** [comment]
Yeah that or the initramfs is not being rebuilt to incorporate the changes on real root and then there are bugs which might be something like the tool rebuilds the initramfs and removes the kernel command line from the boot loader config ( as it should ) but nm in initrd ignores that/those file(s) which might be the case here hence this needs to be tested with dracut 51+.

**Comment by johannbg** [comment]
@superteece is this still an issue?

**Comment by pvalena** [comment]
It seems we've encountered this as well. 

 - Wouldn't `nmcli device disconnect $DEVICE; dracut -f` workaround preserving the state of the device? Is there any way of not activating the device on boot?
 - Alternatively, if some connection is to be used, but with different config, the following:  `nmcli connection modify $CONNECTION connection.autoconnect-priority 10`; and then `dracut -f`?

**Comment by pvalena** [comment]
> It seems we've encountered this as well.
> 
> * Wouldn't `nmcli device disconnect $DEVICE; dracut -f` workaround preserving the state of the device? Is there any way of not activating the device on boot?

No info about connections is probably copied to the initramfs, so `dracut -f` seems to be redundant. But it would be nice to have the state preserved across boots.

> * Alternatively, if some connection is to be used, but with different config, the following:  `nmcli connection modify $CONNECTION connection.autoconnect-priority 10`; and then `dracut -f`?

**Comment by pvalena** [comment]
What I've found is that NetworkManager saves the devices which should not be used/autocreated in `/var/lib/NetworkManager/no-auto-default.state` .... dracut should probably respect that.

**Comment by pvalena** [comment]
* Meaning the file should probably copied and read by `/usr/libexec/nm-initrd-generator`

**Comment by bengal** [comment]
Hi, when NM runs in initrd, it saves its state in `/run/NetworkManager`. After switch root, the new NM instance reads the content of that directory (which includes initrd-generated connection profiles and a information on what profile was active on what device), so that the state is propagated from initrd to real root.

The reason to do this propagation is that there are cases (e.g. boot from a network share) requiring that the network is not changed on switch root. At the moment there isn't any automatic mechanism in NM to avoid that propagation, and the only workaround is to either:
- create a new configuration in real root specifying `keep-configuration=no` and `allowed-connections=except:origin:nm-initrd-generator` for the device. See NM commit https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/commit/bace14fe1f374db26e49e4e7d61d2fbfce4241cc .
- add a new service in real root that runs before NM, clears the state directory and flushes interfaces.

Since those solutions are inconvenient, probably there should be a kernel command line argument that makes the NM dracut module drop all the state before switch root (and possibly, set `keep-configuration=no`). What do you think?

**Comment by bengal** [comment]
CC @thom311

**Comment by thom311** [comment]
> Since those solutions are inconvenient, probably there should be a kernel command line argument that makes the NM dracut module drop all the state before switch root (and possibly, set keep-configuration=no). What do you think?

I am not against a `keep-configuration=no` boot option, but I wonder how useful it is.

For `keep-configuration=no` to make sense, you probably anyway need to configure the real-root. It's not that you could boot an machine the first time and it's gonna work (is it?). If you already configure the real-root, then `keep-configuration=no` + `allowed-connections=except:origin:nm-initrd-generator` seems to be the solution (not merely a workaround). Doesn't seem inconvenient to me...

**Comment by pvalena** [comment]
> Hi, when NM runs in initrd, it saves its state in `/run/NetworkManager`. After switch root, the new NM instance reads the content of that directory (which includes initrd-generated connection profiles and a information on what profile was active on what device), so that the state is propagated from initrd to real root.

IAFAIU, this is is caused  about configuration NOT propagating the other way. F.e. when 'generating' the connections in the intrd, interfaces which were supposed to be disabled are re-enabled, going against the configuration saved in the `root`. Is there a right way to propagate such configuration into initrd?

**Comment by bengal** [comment]
> IAFAIU, this is is caused about configuration NOT propagating the other way. F.e. when 'generating' the connections in the initrd, interfaces which were supposed to be disabled are re-enabled, going against the configuration saved in the `root`. Is there a right way to propagate such configuration into initrd?

Normally, no interface is supposed to be configured in the initrd. Only if there is a kernel command line specifying some network configuration, then nm-initrd-generator creates the corresponding connection profiles. Therefore, I don't understand what you mean by "propagating" the state from real root to initrd: in initrd the only state comes from the kernel command line.

Actually, that's not completely true, users should be able to include connection profiles in the initrd image and they will be honored. What I meant is that propagating the disabled state from real root doesn't make much sense because in initrd interfaces are disabled by default.

Ok, reading the thread again, probably you are referring to the scenario where the command line contains only `rd.neednet`. In absence of other networking options, the generator considers `rd.neednet` in the same way as `ip=dhcp` and generates a wildcard connection profile that does DHCP on all available interfaces.

If activating all devices bothers you and you are willing to rebuild the initrd image, you can decide which interfaces to activate by adding a built-in kernel command line that specifies exactly what to configure (for example, `ip=enp1s0:dhcp` or `ip=[2001:db8::2]:::56::enp2s0`). There is no automatic mechanism to include inside the initrd information about which interfaces should be disabled.

**Comment by bengal** [comment]
> For `keep-configuration=no` to make sense, you probably anyway need to configure the real-root. It's not that you could boot an machine the first time and it's gonna work (is it?). If you already configure the real-root, then `keep-configuration=no` + `allowed-connections=except:origin:nm-initrd-generator` seems to be the solution (not merely a workaround). Doesn't seem inconvenient to me...

Right, you probably need some configuration in the real root. 

On the other hand, having to modify the persistent configuration on the real root to say "override what was configured from initrd" seems backwards to me. The persistent configuration should stay as it is, and the kernel command line that adds the initrd configuration should say "this doesn't propagate to real root".

That said, I don't have specific use cases in mind where one or the other way are clearly best.

**Comment by maltezacharias** [comment]
> I am not against a `keep-configuration=no` boot option, but I wonder how useful it is.


To give some more input, I just ran into this problem. It is probably not often (in my case I use a complicated Bridge, VLAN, etc. configuration in the real system, but only need one VLAN in the initrd to unlock the drive). I think an additional parameter would be a lot easier to figure out, I looked in `man dracut.cmdline` for a solution and didn't find any. 

I think it is counter-intuitive, that when you start to use a network-configuration in your initrd, the configuration made before is partially rendered inactive. I think a parameter like keep-config could be a good idea, but in the end I solved it like @bengal  suggested in the following way:

```
# cat /etc/NetworkManager/conf.d/00-system.conf
[device]
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```

This also _feels_ like the right solution to me, *but* it was very hard to find for me. Maybe it could be a good idea to add this to the dracut.cmdline manpage near or under rd.neednet / ip parameters. 
Sorry for reviving this old issue, I hope there were no docs I overlooked that were created in the meantime

**Comment by bengal** [comment]
> This also _feels_ like the right solution to me, _but_ it was very hard to find for me. Maybe it could be a good idea to add this to the dracut.cmdline manpage near or under rd.neednet / ip parameters. Sorry for reviving this old issue, I hope there were no docs I overlooked that were created in the meantime

`dracut.cmdline` describes the options supported by all the network modules. Since this configuration is NM-specific, it should probably go into the `nm-initrd-generator` man page, what do you think?

**Comment by baryluk** [comment]
@maltezacharias We had a similar issue.

We use clevis (nbde, with tang plugin), but also a static IP assignment on kernel command line (using dracut's extra cmdline in the dracut config) - with mac assignment to interfaces, `bond0` forming, and ip settings set on `bond0` + nameservers. And separately static nm configs in the root. After a boot, we would see both ones from initrd (in `/run/NetworkManager/`) and ones from root (in `/etc/NetworkManager/`). Fortunately ones from the root being higher priority, the connectivity is not broken. But the ones from initrd still showing up in `nmcli con show`, is very confusing.

Also for completeness we also had this:

```
# cat /etc/NetworkManager/conf.d/50-foobar.conf
[main]
no-auto-default=*
```

I just tried

```
# cat /etc/NetworkManager/conf.d/00-system.conf
[device]
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```

but still the same:

```
[root@foobar ~]# nmcli con show
NAME   UUID                                  TYPE      DEVICE 
bond0  4d515eaf-5a30-4caa-90f1-e97ff3d60b7a  bond      bond0     green
em1    30626f2c-2865-4db0-9fa7-8d21c7373c8c  ethernet  em1       green
em2    bfeaf765-0123-42e2-adf7-10cc781493da  ethernet  em2       green
lo     ff78282d-7aca-4739-84cc-3e7e9b62d6ce  loopback  lo        dark green
bond0  20e87b42-92e0-4b88-ba3e-ca897718c0cd  bond      --        white  (random uuid on every reboot)
em1    f15ef34f-f430-422b-afec-07d27b937ae8  ethernet  --        white  (ditto)
em2    a888ea7e-5d9a-44e7-aea5-838613c314c4  ethernet  --        white  (ditto)
```


(NetworkManager startup log does show these old interface do transition `unmanaged -> unavailable (reason 'managed', managed-type: 'external')`, and a bit later the proper ones are activated, including interface renaming)


Here are the configs:


```
[root@foobar ~]# grep -r -E 'origin|uuid|method' /run/NetworkManager/system-connections/
/run/NetworkManager/system-connections/lo.nmconnection:uuid=ff78282d-7aca-4739-84cc-3e7e9b62d6ce
/run/NetworkManager/system-connections/lo.nmconnection:method=manual
/run/NetworkManager/system-connections/lo.nmconnection:method=manual
/run/NetworkManager/system-connections/em2.nmconnection:uuid=a888ea7e-5d9a-44e7-aea5-838613c314c4
/run/NetworkManager/system-connections/em2.nmconnection:org.freedesktop.NetworkManager.origin=nm-initrd-generator
/run/NetworkManager/system-connections/em1.nmconnection:uuid=f15ef34f-f430-422b-afec-07d27b937ae8
/run/NetworkManager/system-connections/em1.nmconnection:org.freedesktop.NetworkManager.origin=nm-initrd-generator
/run/NetworkManager/system-connections/bond0.nmconnection:uuid=20e87b42-92e0-4b88-ba3e-ca897718c0cd
/run/NetworkManager/system-connections/bond0.nmconnection:method=manual
/run/NetworkManager/system-connections/bond0.nmconnection:method=disabled  # ipv6
/run/NetworkManager/system-connections/bond0.nmconnection:org.freedesktop.NetworkManager.origin=nm-initrd-generator
[root@foobar ~]# 


[root@foobar ~]# grep -r -E 'origin|uuid|method' /etc/NetworkManager/system-connections/
/etc/NetworkManager/system-connections/bond0.nmconnection:uuid=4d515eaf-5a30-4caa-90f1-e97ff3d60b7a
/etc/NetworkManager/system-connections/bond0.nmconnection:method=manual
/etc/NetworkManager/system-connections/bond0.nmconnection:method=ignore  # ipv6
/etc/NetworkManager/system-connections/em1.nmconnection:uuid=30626f2c-2865-4db0-9fa7-8d21c7373c8c
/etc/NetworkManager/system-connections/em2.nmconnection:uuid=bfeaf765-0123-42e2-adf7-10cc781493da
[root@foobar ~]# 


[root@foobar ~]# grep -r . /etc/NetworkManager/conf.d/
/etc/NetworkManager/conf.d/00-system.conf:[device]
/etc/NetworkManager/conf.d/00-system.conf:keep-configuration=no
/etc/NetworkManager/conf.d/00-system.conf:allowed-connections=except:origin:nm-initrd-generator
/etc/NetworkManager/conf.d/50-foobar.conf:[main]
/etc/NetworkManager/conf.d/50-foobar.conf:no-auto-default=*
[root@foobar ~]# 
```



```
[root@foobar ~]# NetworkManager --print-config
# NetworkManager configuration: /etc/NetworkManager/NetworkManager.conf, /usr/lib/NetworkManager/conf.d/{99-nvme-nbft-no-ignore-carrier.conf,tensor-no-auto-default.conf}, /run/NetworkManager/conf.d/15-carrier-timeout.conf, /etc/NetworkManager/conf.d/{00-system.conf,50-foobar.conf}

[main]
# plugins=
# rc-manager=auto
# migrate-ifcfg-rh=false
# auth-polkit=true
# dhcp=internal
# iwd-config-path=
no-auto-default=*
configure-and-quit=no

[logging]
# backend=journal
# audit=false

[device]
# wifi.backend=wpa_supplicant
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator

[device-nbft-no-ignore-carrier]
match-device=interface-name:nbft*
ignore-carrier=no

# no-auto-default file "/var/lib/NetworkManager/no-auto-default.state"
```


And dracut configs:

```
[root@foobar dracut.conf.d]# grep -r .
network.conf:omit_dracutmodules+=" ifcfg "
nbde.conf:add_drivers+=" bonding "
nbde.conf:kernel_cmdline+=" rd.neednet=1 "
nbde.conf:kernel_cmdline+=" ifname=em1:aa:bb:cf:11:22:30 "
nbde.conf:kernel_cmdline+=" ifname=em2:aa:bb:cf:11:22:31 "
nbde.conf:kernel_cmdline+=" bond=bond0:em1,em2:mode=802.3ad,miimon=100,lacp_rate=fast,updelay=200,downdelay=200,xmit_hash_policy=layer3+4 "
nbde.conf:kernel_cmdline+=" ip=1.2.3.7::1.2.3.254:255.255.255.0:foobar.example.com:bond0:none "
nbde.conf:kernel_cmdline+=" nameserver=1.2.3.4 "
nbde.conf:kernel_cmdline+=" rd.luks.uuid=8a547de7-e9ef-436d-9e73-1b5617e95bb1 "
nbde.conf:kernel_cmdline+=" rd.md.uuid=foo:bar:baz "
nbde.conf:kernel_cmdline+=" rd.md.uuid=foo:bar:baz "
nbde.conf:kernel_cmdline+=" rd.md.uuid=foo:bar:baz "
nbde.conf:add_dracutmodules+=" network-manager "
nbde.conf:add_device+=" /dev/disk/by-uuid/8a547de7-e9ef-436d-9e73-1b5617e95bb1 "

```


Version info:

CentOS 9, 

```
[root@foobar ~]# NetworkManager --version
1.53.1-1.el9
[root@foobar ~]# dracut --version
dracut 057-87.git20250311.el9
```

**Comment by paulraines68** [comment]
This is causing me a worrying issue now on a new Rocky 9 server I am trying to configure with full encryption and using clevis/Tang to auto decrypt on boot.

I realize now after finding this issue on this Rocky 9 server that two previous Rocky 8 VMs I setup with full encryption and clevis/Tang have the same issue but since they are doing DHCP I did not notice till now becuase network just "worked".   On those VMs, running 'nmcli conn' shows:

```
NAME              UUID                                  TYPE      DEVICE
Wired Connection  50c02611-2af2-4203-a712-96a5ba76b530  ethernet  enp1s0 (green)
lo                c627f5c9-b65a-4e80-be5d-08bdea91ee3f  loopback  lo
enp1s0            5ce3012e-ffa8-4b44-bb93-21f6aee93f7a  ethernet  -- (grey)

```
Where enp1s0 was the working connection using DHCP before I setup clevis/Tang and Wired Connection is this new thing from nowhere I now get, but it is doing DHCP and works so the Linux OS has network the same essentially as enp1s0 used to

However, with the new server I am trying to do static IP and eventually want bonding like @baryluk above.    Before setting up clevis/Tang on the box my connections where like:

```
NAME         UUID                                  TYPE      DEVICE
lo           b0ca004b-2548-4442-a6b2-6e3bcef6559a  loopback  lo
eno12399np0  d3d2f008-fe34-4853-8517-456e77c0d1ff  ethernet  eno12399np0 (green)
eno12409np1  245aa6f4-3341-4b0f-ba6c-7717d112a01c  ethernet  --
```

No bonding setup yet so only eno12399np0 was setup with the static IP.  eno12409np1 was left unconfigured and stay down at boot. 

After Installing clevis/Tang it did fine auto decrypting but once it booted to the OS I could not reach it at the static IP I gave.  I logged in on its virtual console via BMC and found a situation like (I am making up UIDs as I did not record them):

```
NAME              UUID                                  TYPE      DEVICE
Wired Connection  50c02611-2af2-4203-a712-96a5ba76b530  ethernet  eno12399np0 (green)
Wired Connection  50c02611-2af2-4203-a712-96a5ba76b530  ethernet  eno12409np1 (green)
lo           b0ca004b-2548-4442-a6b2-6e3bcef6559a  loopback  lo
eno12399np0  d3d2f008-fe34-4853-8517-456e77c0d1ff  ethernet  --   (grey)
eno12409np1  245aa6f4-3341-4b0f-ba6c-7717d112a01c  ethernet  --  (grey)
```

After running 'nmcli conn delete "Wired Connection"' suddenly eno12399np0 took over and I was back online at my static IP.

So after web searching for this issue and finding other complaining about this issue but no one listing any definitive solution I started playing around.  After an initial try using:

```
# cat /etc/dracut.conf.d/clevis.conf
hostonly_cmdline=yes
# cat /etc/dracut.conf.d/network.conf
omit_dracutmodules+=" ifcfg "
add_dracutmodules+=" network-manager "
kernel_cmdline+=" rd.neednet=1 "
kernel_cmdline+=" ifname=eno12399np0:6C:92:CF:E7:3C:F0 "
kernel_cmdline+=" ip=172.27.70.35::172.27.70.1:255.255.255.0:lcnsrv07:eno12399np0:none "
kernel_cmdline+=" nameserver=172.21.21.45 "
# cat /etc/NetworkManager/conf.d/00-system.conf
[device]
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```

also putting the kernel cmdline stuff in /etc/default/grub and remaking grub,  on boot I would have no working network at all.  There was some kind of conflict between the eno12399np0 I already had defined in NetworkManager and the one this creates during the boot dracut initrd process.   I ended up just deleting the /etc/NetworkManager/system-connections/eno12399np0.nmconnection file and rebooting.  This booted up a system with eno12399np0 connection doing DHCP and a new /etc/NetworkManager/system-connections/eno12399np0.nmconnection file created.  I modifed that file which had a new UID now to do the statci IP I wanted.  Then rebooted and now I at least have a working network on the IP I want without intervention at boot that looks like this:

```
NAME         UUID                                  TYPE      DEVICE
eno12399np0  e88ba15c-d02c-4757-9518-241f297144dc  ethernet  eno12399np0 (green)
lo           994b1275-accc-452d-9c68-f6200f73b1b2  loopback  lo
eno12399np0  e231cf85-fe47-4bae-984f-921511b6fc78  ethernet  -- (grey)
eno12409np1  245aa6f4-3341-4b0f-ba6c-7717d112a01c  ethernet  --
```

These UUIDS are now correct since this is what I have now.  The top green line is the UUID in the /etc/NetworkManager/system-connections/eno12399np0.nmconnection file so I have no idea why that grey eno12399np0 is even defined or where its UUID comes from.

So I am still uncertain what the "correct" way to do things here and how do I get a bond working when I want to configure to do that.

**Comment by yrro** [comment]
> I have no idea why that grey eno12399np0 is even defined or where its UUID comes from.

Try `nmcli -f name,state,type,device,filename con` 

If you see a profile living in `/run/NetworkManager` then it's likely to be the connection profile created during initrd-time.

If you modify such a profile _and then save it_ then it will be persisted to `/etc/NetworkManager/system-connections` and hang around until you delete it.

> So I am still uncertain what the "correct" way to do things here and how do I get a bond working when I want to configure to do that.

If you configure networking via kernel parameters then the connection created at initrd-time will remain active and and any connection profiles that might exist on the final system (i.e,. in `/etc/NetworkManager/system-connections`) will be ignored.

One approach is to configure _all_ networking options via kernel parameters. So you'd need to add an appropriate `bond=` parameter. Then you can ignore (and should probably remove, in order to prevent future confusion) any connection profiles configured in the root filesystem.

The other approach is to regard the initrd-time connection profile as a temporary one that is only used for bootstrapping the system to the point where the root filesystem can be mounted, and configure NetworkManager to disconnect the initrd-time connection profile and connect a connection profile stored within the root filesystem when it starts up. That's what the config examples in prior comments that use the `allowed-connections=` option do.

**Comment by paulraines68** [comment]
> The other approach is to regard the initrd-time connection profile as a temporary one that is only used for bootstrapping the system to the point where the root filesystem can be mounted, and configure NetworkManager to disconnect the initrd-time connection profile and connect a connection profile stored within the root filesystem when it starts up. That's what the config examples in prior comments that use the `allowed-connections=` option do.

I do have 

```
# cat /etc/NetworkManager/conf.d/00-system.conf
[device]
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```

but this seemed to have no effect in preventing the dracut initrd defined networks from being there at the OS boot 

But I must admit I have no idea what those values for allowed-connections really mean.  Maybe they need to be something else.

**Comment by yrro** [comment]
The dracut-defined connection profiles will still exist, but they should be inactive. If you look at their filename property you should see they are located within `/run/NetworkManager`; they are transient/not stored persistently. If you look inside the files you should see they have `org.freedesktop.NetworkManager.origin=nm-initrd-generator` which identifies them as a connection to be disallowed by the NM config snippet.

**Comment by paulraines68** [comment]
Okay, I see now

```
NAME         STATE      TYPE      DEVICE       FILENAME                                                  
eno12399np0  activated  ethernet  eno12399np0  /etc/NetworkManager/system-connections/eno12399np0.nmconnec (green)
lo           activated  loopback  lo           /run/NetworkManager/system-connections/lo.nmconnection     
eno12399np0  --         ethernet  --           /run/NetworkManager/system-connections/eno12399np0.nmconnec
eno12409np1  --         ethernet  --           /etc/NetworkManager/system-connections/eno12409np1.nmconnec
```

Still not sure why I had delete the previous working eno12399np0 connection to get it to work but I guess something in the process made the UUID change.

**Comment by paulraines68** [comment]
Trying to get bonding setup I could not get it to work where it discarded the dracut defined network and used what I had in /etc/NetworkManager/system-connections.  It says my devices are invalid even though they worked fine before.  I just gave up on that method.

Instead I removed the NetworkManager settings

```
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```

to just keep what I had defined for dracut in /etc/dracut.conf.d/network.conf and in /etc/default/grub (not clear to me if I need both).

This works for me now since I discovered one can configure the MTU in the kernel ip= line with

ip=172.18.19.38::172.18.19.1:255.255.255.0:lcnsrv08:bond0:none:9000

but if I need to configure any other less common properties, I am not sure how I will do it.  Maybe just setup a boot script to modify the dracut-made connection in systemd after boot.

**Comment by jorti** [comment]
I faced the same issue and I resolved it by adding this configuration. Note that I've added a `match-device` option which is something I don't see in the previous examples:

```
# cat /etc/NetworkManager/conf.d/00-custom.conf 
[device]
match-device=interface-name:enp88s0
keep-configuration=no
allowed-connections=except:origin:nm-initrd-generator
```
